✨ Adapter for Windows

acabarbaye opened the pull request:
### Description

New adapter for Windows.

Caveat:
Some changes may be needed for ST boards as the serial port does not seem to be retrieved properly since I updated the USB drivers with Mbed Studio. Would be worth testing on a fresh Windows machine.

### Test Coverage

- [X]  This change is covered by existing or additional automated tests.
- [X]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

rwalton-arm commented on mbed_devices/_internal/utils/python_helpers.py:
```diff
@@ -0,0 +1,23 @@
+    if is_python_more_recent(7):
+        return namedtuple(typename=typename, field_names=field_names, defaults=defaults)  # type:ignore
+    else:
+        tuple: NamedTuple = namedtuple(typename=typename, field_names=field_names)  # type:ignore
```
I wish we could use `dataclasses` instead of `NamedTuples`. Unfortunately they're only part of the stdlib in python 3.7+

muchzill4 commented on mbed_devices/_internal/windows/component_descriptor.py:
```diff
@@ -0,0 +1,123 @@
+        return None
+
+    @property
+    def is_undefined(self) -> bool:
```
This is a hidden `None` check. I think we should go with pattern similar to `Candidate`, where _empty_ instances should not event exist. Otherwise the burden is on whoever uses the instance to check if it's empty or not. Otherwise, what's the use of `empty` value?

rwalton-arm commented on mbed_devices/_internal/windows/component_descriptor.py:
```diff
@@ -0,0 +1,123 @@
+        return self._class_parameters()
+
+    @classmethod
+    def _class_parameters(cls) -> List[str]:
```
This shouldn't be a `@classmethod` I'd expect it to return an instance of `cls` where this returns a list of class attributes.

rwalton-arm commented on mbed_devices/_internal/windows/component_descriptor.py:
```diff
@@ -0,0 +1,123 @@
+    @abstractmethod
+    def device_id(self) -> str:
+        """Returns the device ID field."""
+        pass
```
Don't need `pass` here. The docstring should be enough.

rwalton-arm commented on mbed_devices/_internal/windows/component_descriptor.py:
```diff
@@ -0,0 +1,123 @@
+    @property
+    def filter(self) -> Optional[str]:
+        """Filter applied on the list of elements."""
+        return None
```
?

acabarbaye replied:
Tried to improve the documentation

rwalton-arm commented on mbed_devices/_internal/windows/component_descriptor.py:
```diff
@@ -0,0 +1,123 @@
+        """Determines whether the structure is undefined or not i.e. none of the fields are actually defined."""
+        undefined_values = []
+        for param in self.parameters:
+            value = self.__getattribute__(param)
```
I would prefer using `getattr` over calling `__getattribute__` directly.

rwalton-arm commented on mbed_devices/_internal/windows/component_descriptor.py:
```diff
@@ -0,0 +1,123 @@
+    @staticmethod
+    def _read_cdispatch_field(win32_element: Any, key: str) -> Any:
+        try:
+            return getattr(win32_element, key)
```
Why not just do `win32_element.key` if we're catching the `AttributeError` anyway?

rwalton-arm replied:
Oh, "key" is a string.

rwalton-arm commented on mbed_devices/_internal/windows/component_descriptor.py:
```diff
@@ -0,0 +1,123 @@
+logger = logging.getLogger(__name__)
+
+
+class ComponentDescriptor(ABC):
```
I'm really struggling to understand what this object does and why we need all of this complexity. It seems like some sort of abstract factory class which instantiates a `NamedTuple` in a bunch of different ways, with some unexpected uses of `@classmethod`. Do we really need to make this so complex? I'd prefer some simple factory functions over this abstract factory (if that's what this is).

acabarbaye replied:
I radically changed the code to make it simpler to follow. Hopefully, I succeeded.

rwalton-arm commented on mbed_devices/_internal/windows/disk_aggregation.py:
```diff
@@ -0,0 +1,189 @@
+    @classmethod
+    def iter_all(cls) -> Generator["ComponentDescriptor", None, None]:
+        """Gets a generator over all elements currently registered in the system."""
+        yield from ()
```
I have a few questions about this method:
Is this `yield from ()` a TODO or something? I don't understand what's going on here.
I don't like the name `iter_all` what does "all" mean?
Why is this a `@classmethod`?

rwalton-arm commented on mbed_devices/_internal/windows/disk_aggregation.py:
```diff
@@ -0,0 +1,189 @@
+        yield from ()
+
+
+class DataAggregator(ABC):
```
This name seems a bit ambiguous to me, what kind of data are we aggregating? The docstring mentions "disk data", what is that? I suggest using more descriptive names and providing more context in the docstrings.

rwalton-arm commented on mbed_devices/_internal/windows/disk_aggregation.py:
```diff
@@ -0,0 +1,189 @@
+    """Aggregator of disk data."""
+
+    def __init__(self, physical_disks: dict, partition_disks: dict, logical_partition_relationships: dict) -> None:
+        """Constructor."""
```
`__new__` is the constructor, this is the initialiser.

rwalton-arm commented on mbed_devices/_internal/windows/disk_aggregation.py:
```diff
@@ -0,0 +1,189 @@
+        self._partition_disks = partition_disks
+        self._logical_partition_relationships = logical_partition_relationships
+
+    def _get_corresponding_physical_disk(self, partition: DiskPartition) -> DiskDrive:
```
I'm not sure it's worth having this single line in a function when there's only a single caller that I can see, the actual dictionary lookup is descriptive enough for me to understand what's happening.

muchzill4 commented on Pipfile:
```diff
@@ -18,3 +18,6 @@ mbed-tools-ci-scripts = "*"
 allow_prereleases = true
+
+[packages]
+mbed-devices = {path = "."}
```
Wild dependency appears. 🧐

acabarbaye commented on mbed_devices/_internal/windows/usb_device_descriptor.py:
```diff
@@ -0,0 +1,154 @@
+
+
+class UsbIdentifier(
+    named_tuple_with_defaults(  # type:ignore
```
it does not set all the values to None. Just the ones which were not provided. The reason this is done is that there is no rule about what elements are present in the DeviceID and what isn't

acabarbaye commented:
Due to the amount of review comments, I refactored drastically the code and will try to answer most change requests in this comment.
1. The warning printing when run as part of pytest come form pywin32 dependency which I use to make the calls to the Windows system. I will not be able to change this.
2. I reduced the number of `#type: ignore` to a handful and limited them to where I could not find any better ways.
3. I'd prefer to keep the definitions of the different Win32 class as close as their definition in MSDN so that it is easy to follow and maintain in the future. It also makes it easier to debug.
4.5. I refactored the code so that it uses composition rather than multiple inheritance. Hopefully I removed the evil although it may remain in the details. ComponentDescriptor is now much smaller and I removed all the code trying to eliminate duplicates.

muchzill4 commented:
@acabarbaye I'm sorry, but I'm not willing to review this again. I still see things I'm unhappy about, but I don't find it productive to work against such big diff.

I think we should revisit the approach we take with regards to pull requests, especially that `maintainability` is part of our motto. ;)
